### ISSUE TITLE
built-in fetch: constant polling causes memory leak (Android)

### ISSUE THREAD

**Issue by vitto-moz** [facebook/react-native#29065]
## Description
We have the kiosk-like app built for android only. The application makes constant polling for the data once per minute and the response (JSON part) would be quite massive. We've noticed a memory leak every time on built-in **fetch** invocation. Due to the long-running nature of the app (about 20 hours per day), we have crashes and lags because of the constant increasing of memory consumption.

## React Native version:
System:
    OS: macOS Mojave 10.14.6
    CPU: (8) x64 Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz
    Memory: 162.09 MB / 16.00 GB
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 10.15.3 - /usr/local/bin/node
    Yarn: 1.19.0 - /usr/local/bin/yarn
    npm: 6.4.1 - /usr/local/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.7.5 - /usr/local/bin/pod
  SDKs:
    iOS SDK:
      Platforms: iOS 12.2, macOS 10.14, tvOS 12.2, watchOS 5.2
    Android SDK:
      API Levels: 23, 25, 26, 27, 28
      Build Tools: 23.0.1, 25.0.0, 25.0.2, 26.0.1, 26.0.2, 26.0.3, 28.0.3
      System Images: android-22 | Google APIs Intel x86 Atom, android-25 | Intel x86 Atom, android-27 | Google Play Intel x86 Atom, android-28 | Google Play Intel x86 Atom, android-29 | Google Play Intel x86 Atom, android-R | Google APIs Intel x86 Atom
      Android NDK: Not Found
  IDEs:
    Android Studio: 3.4 AI-183.5429.30.34.5452501
    Xcode: 10.2.1/10E1001 - /usr/bin/xcodebuild
  Languages:
    Java: 1.8.0_211 - /usr/bin/javac
    Python: 2.7.10 - /usr/bin/python
  npmPackages:
    @react-native-community/cli: Not Found
    react: 16.11.0 => 16.11.0
    react-native: 0.62.2 => 0.62.2
  npmGlobalPackages:
    *react-native*: Not Found

## Steps To Reproduce

1. build a simple app with polling, which uses a **fetch**. To get the result faster use the endpoint which returns massive JSON data and increase polling frequency (once per 10 seconds for example)
2. run application and use "adb shell dumpsys meminfo com.yourapp" to get info about application memory consumption (run the command sequentially once per minute and observe constant increasing of memory consumption)

## Expected Results
**fetch** should release memory once there is no link to the parsed JSON response (response.json()) - in simple words, garbage collected.

## Snack, code example, screenshot, or link to a repository:

```
  function startUpdating() {
      const delay = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));
      while (true) {
        getMoviesFromApi()
        await delay(10000);
      }
    }

function getMoviesFromApi() {
  return fetch('https://reactnative.dev/movies.json') // endpoint which returns massive data
    .then((response) => response.json())
    .then((json) => {
      return json.movies;
    })
    .catch((error) => {
      console.error(error);
    });
}

startUpdating()
```

**Comment by vitto-moz** [comment]
Sorry, I had to change the issue explanation again as I had realized it's **fetch** call itself who cause this behavior with constantly rising memory consumption. 

btw, the issue might be related to this one
https://github.com/facebook/react-native/issues/23801

**Comment by vitto-moz** [comment]
The issue has been resolved by **moving the request itself on the android native side** and getting results back to the JS layer through the bridge. Important to mention that a timer functionality is still on the JS layer, which confirms that the timer is not a source of the issue.

so, my current implementation looks kind of this:
1) JS: trigger the react method on the native side by timer through the bridge 
2) Native: execute the request and receive the response
3) Native: invoke the callback with a response
4) JS: getting the response in a callback and handling it

**Comment by vitto-moz** [comment]
I think this issue really requires the attention of the community. 
it's important to fix it - to have all long-running apps made with react-native use less memory

**Comment by mamartin** [comment]
we had similar problem with memory leak while fetching. I noticed that even new empty app (RN54) is constantly increasing used memory if response.json() is omitted from the example below (setInterval is called only once)

`    setInterval(() => {
      fetch('https://postman-echo.com/bytes/1/mb?type=json', {
        credentials: 'omit',
        headers: {
          accept: '*/*',
          'content-type': 'application/json',
        },
        referrerPolicy: 'no-referrer-when-downgrade',
        method: 'GET',
        mode: 'no-cors',
      })
        .then((response) => {
          return response.json();
        })
        .then((response) => {
          console.log('===== response', response); // eslint-disable-line no-console
        })
        .catch((e) => {
          console.log('===== error', e); // eslint-disable-line no-console
        });
    }, 3000);`

with `esponse.json()` its not leaking on RN 64, but its leaking in RN 59.

following image compares used memory in RN63 vs RN59 in our test scenario:

![image](https://user-images.githubusercontent.com/940559/128844933-96782432-4ecc-4e84-926e-00b5087f3d88.png)

another thing that helped us was upgrading from Apollo2 to Apollo3.

**Comment by hezhii** [comment]
Is there any progress regarding this issue? This issue still persists in react-native version 0.71.7.

```javascript
useEffect(() => {
    (async () => {
      for (let i = 0; i < 100000; i++) {
        console.log('start');
        let timer: any;
        try {
          const abortController = new AbortController();
          timer = setTimeout(() => {
            abortController.abort();
          }, 120 * 1000);
          const res = await fetch(
            'https://postman-echo.com/bytes/1/mb?type=json',
            {
              signal: abortController.signal,
            },
          );
          const json = await res.json();
        } catch (err) {
          console.log('err:', err);
        } finally {
          clearTimeout(timer);
        }
        console.log('finished');
      }
    })();
  }, []);
```

**Comment by hezhii** [comment]
Is there any progress regarding this issue? This issue still persists in react-native version 0.71.7.

**Comment by varpstar** [comment]
Same issue here

**Comment by ipnet-me** [comment]
I have similar problem, I download and save around 2400 binary files to android storage, but after some time app is crash with memory error text

**Comment by gavrilikhin-d** [comment]
This happens with fetching images too.

**Comment by nero2009** [comment]
This is still an issue in "react-native": "0.74.5"

**Comment by nero2009** [comment]
Here is a reproduction https://github.com/nero2009/fetch-memory-leak

**Comment by migueldaipre** [comment]
> Here is a reproduction https://github.com/nero2009/fetch-memory-leak

Hey @nero2009, this issue is closed. 

Please create a new issue if the memory leak still happens.
